test/cpp/naming/naming_end2end_test.cc

```diff
@@ -0,0 +1,297 @@
+  args->channel_args = NULL;
+}
+
+void args_finish(grpc_exec_ctx *exec_ctx, args_struct *args) {
+  GPR_ASSERT(gpr_event_wait(&args->ev, test_deadline()));
+  grpc_pollset_set_del_pollset(exec_ctx, args->pollset_set, args->pollset);
+  grpc_pollset_set_destroy(exec_ctx, args->pollset_set);
+  grpc_closure do_nothing_cb;
+  GRPC_CLOSURE_INIT(&do_nothing_cb, do_nothing, NULL,
+                    grpc_schedule_on_exec_ctx);
+  grpc_pollset_shutdown(exec_ctx, args->pollset, &do_nothing_cb);
+  // exec_ctx needs to be flushed before calling grpc_pollset_destroy()
+  grpc_channel_args_destroy(exec_ctx, args->channel_args);
+  grpc_exec_ctx_flush(exec_ctx);
+  grpc_pollset_destroy(exec_ctx, args->pollset);
+  gpr_free(args->pollset);
+  GRPC_COMBINER_UNREF(exec_ctx, args->lock, NULL);
+}
+
+static gpr_timespec n_sec_deadline(int seconds) {
+  return gpr_time_add(gpr_now(GPR_CLOCK_REALTIME),
+                      gpr_time_from_seconds(seconds, GPR_TIMESPAN));
+}
+
+static void poll_pollset_until_request_done(args_struct *args) {
+  gpr_timespec deadline = n_sec_deadline(10);
+  while (true) {
+    bool done = gpr_atm_acq_load(&args->done_atm) != 0;
+    if (done) {
+      break;
+    }
+    gpr_timespec time_left =
+        gpr_time_sub(deadline, gpr_now(GPR_CLOCK_REALTIME));
+    gpr_log(GPR_DEBUG, "done=%d, time_left=%" PRId64 ".%09d", done,
+            time_left.tv_sec, time_left.tv_nsec);
+    GPR_ASSERT(gpr_time_cmp(time_left, gpr_time_0(GPR_TIMESPAN)) >= 0);
+    grpc_pollset_worker *worker = NULL;
+    grpc_exec_ctx exec_ctx = GRPC_EXEC_CTX_INIT;
+    gpr_mu_lock(args->mu);
+    GRPC_LOG_IF_ERROR(
+        "pollset_work",
+        grpc_pollset_work(&exec_ctx, args->pollset, &worker,
+                          gpr_now(GPR_CLOCK_REALTIME), n_sec_deadline(1)));
+    gpr_mu_unlock(args->mu);
+    grpc_exec_ctx_finish(&exec_ctx);
+  }
+  gpr_event_set(&args->ev, (void *)1);
+}
+
+static void check_service_config_result_locked(grpc_channel_args *channel_args,
+                                               args_struct *args) {
+  const grpc_arg *service_config_arg =
+      grpc_channel_args_find(channel_args, GRPC_ARG_SERVICE_CONFIG);
+  if (args->expected_service_config_string != NULL &&
+      strlen(args->expected_service_config_string) > 0) {
+    GPR_ASSERT(service_config_arg != NULL);
+    GPR_ASSERT(service_config_arg->type == GRPC_ARG_STRING);
+    char *service_config_string = service_config_arg->value.string;
+    if (gpr_stricmp(service_config_string,
+                    args->expected_service_config_string) != 0) {
+      gpr_log(GPR_ERROR, "expected service config string: |%s|",
+              args->expected_service_config_string);
+      gpr_log(GPR_ERROR, "got service config string: |%s|",
+              service_config_string);
+      GPR_ASSERT(0);
+    }
+  } else {
+    GPR_ASSERT(service_config_arg == NULL);
+  }
+}
+
+static void check_channel_arg_srv_result_locked(grpc_exec_ctx *exec_ctx,
```

Comment:
Suggest calling this `check_resolver_result_locked()`, since it's not specific to SRV records.